File: jps-plugin/src/org/jetbrains/jps/cabal/CabalBuilder.java
/*
 * Copyright 2000-2012 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package org.jetbrains.jps.cabal;


import org.jetbrains.annotations.NotNull;
import org.jetbrains.jps.ModuleChunk;
import org.jetbrains.jps.builders.DirtyFilesHolder;
import org.jetbrains.jps.builders.java.JavaSourceRootDescriptor;
import org.jetbrains.jps.incremental.*;
import org.jetbrains.jps.incremental.messages.BuildMessage;
import org.jetbrains.jps.incremental.messages.CompilerMessage;
import org.jetbrains.jps.incremental.messages.ProgressMessage;
import org.jetbrains.jps.model.module.JpsModule;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.*;


public class CabalBuilder extends ModuleLevelBuilder {


    public CabalBuilder() {
        super(BuilderCategory.TRANSLATOR);
    }


    public ExitCode build(final CompileContext context,
                          final ModuleChunk chunk,
                          final DirtyFilesHolder<JavaSourceRootDescriptor, ModuleBuildTarget> dirtyFilesHolder,
                          final OutputConsumer outputConsumer) throws ProjectBuildException {
        try {
            for (JpsModule module : chunk.getModules()) {
                File cabalFile = getCabalFile(module);
                CabalJspInterface cabal = new CabalJspInterface(cabalFile);

                context.processMessage(new CompilerMessage("cabal", BuildMessage.Kind.INFO, "Start configure"));

                Process configureProcess = cabal.configure();

                Iterator<String> processOut = collectOutput(configureProcess);

                while (processOut.hasNext()) {
                    String line = processOut.next();
                    String warningPrefix = "Warning: ";
                    if (line.startsWith(warningPrefix)) {
                        String text = line.substring(warningPrefix.length()) + "\n" + processOut.next();
                        context.processMessage(new CompilerMessage("cabal", BuildMessage.Kind.WARNING, text));
                    } else {
                        context.processMessage(new CompilerMessage("cabal", BuildMessage.Kind.INFO, line));
                    }
                }

                if (configureProcess.waitFor() != 0) {
                    context.processMessage(new CompilerMessage(
                            "cabal",
                            BuildMessage.Kind.ERROR,
                            "configure failed."));
                    return ExitCode.ABORT;
                }
                context.processMessage(new ProgressMessage("Build build"));
                context.processMessage(new CompilerMessage("ghc", BuildMessage.Kind.INFO, "Start build"));
                Process buildProcess = cabal.build();
                processOut = collectOutput(buildProcess);

                while (processOut.hasNext()) {
                    String line = processOut.next();
                    context.processMessage(new CompilerMessage("cabal", BuildMessage.Kind.INFO, line));
                }

                if (buildProcess.waitFor() == 0) {
                    return ExitCode.OK;
                } else {
                    context.processMessage(new CompilerMessage("cabal", BuildMessage.Kind.ERROR, "build errors."));
                }
            }
            return ExitCode.ABORT;
        } catch (InterruptedException e) {
            e.printStackTrace();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return ExitCode.ABORT;
    }

    private Iterator<String> collectOutput(Process process) throws IOException {
        final BufferedReader reader = new BufferedReader (new InputStreamReader(process.getInputStream()));
        return new Iterator<String>() {

            String line = null;

            @Override
            public boolean hasNext() {
                return fetch() != null;
            }

            private String fetch() {
                if (line == null) {
                    try {
                        line = reader.readLine();
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                }
                return line;
            }

            @Override
            public String next() {
                String result = fetch();
                line = null;
                return result;
            }

            @Override
            public void remove() {
                throw new UnsupportedOperationException();
            }
        };
    }

    private File getCabalFile(JpsModule module) {
        String url = module.getContentRootsList().getUrls().get(0);
        try {
            for (File file : new File(new URL(url).getFile()).listFiles()) {
                if (file.getName().endsWith(".cabal")) {
                    return file;
                }
            }
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }
        return null;
    }



    @Override
    public List<String> getCompilableFileExtensions() {
        return Arrays.asList("hs");
    }

    @Override
    public String toString() {
        return getPresentableName();
    }

    @NotNull
    public String getPresentableName() {
        return "Cabal builder";
    }

}

File: jps-plugin/src/org/jetbrains/jps/cabal/CabalBuilder.kt
package org.jetbrains.jps.cabal

import org.jetbrains.jps.ModuleChunk
import org.jetbrains.jps.builders.DirtyFilesHolder
import org.jetbrains.jps.builders.java.JavaSourceRootDescriptor
import org.jetbrains.jps.incremental.*
import org.jetbrains.jps.incremental.messages.BuildMessage
import org.jetbrains.jps.incremental.messages.CompilerMessage
import org.jetbrains.jps.incremental.messages.ProgressMessage
import org.jetbrains.jps.model.module.JpsModule
import java.io.BufferedReader
import java.io.File
import java.io.IOException
import java.io.InputStreamReader
import java.net.MalformedURLException
import java.net.URL
import java.util.*
import org.jetbrains.jps.incremental.ModuleLevelBuilder.ExitCode
import org.jetbrains.jps.incremental.ModuleLevelBuilder.OutputConsumer


public class CabalBuilder() : ModuleLevelBuilder(BuilderCategory.TRANSLATOR) {
    public override fun build(context: CompileContext?,
                              chunk: ModuleChunk?,
                              dirtyFilesHolder: DirtyFilesHolder<JavaSourceRootDescriptor, ModuleBuildTarget>?,
                              outputConsumer: OutputConsumer?): ExitCode {
        try {
            for (module in chunk!!.getModules()!!) {
                val cabalFile = getCabalFile(module)
                val cabal = CabalJspInterface(cabalFile!!)
                if (runConfinguration(context!!, cabal))
                    return ModuleLevelBuilder.ExitCode.ABORT

                if (runBuild(context, cabal))
                    return ModuleLevelBuilder.ExitCode.ABORT

            }
            return ModuleLevelBuilder.ExitCode.OK
        }
        catch (e: InterruptedException) {
            e.printStackTrace()
        }
        catch (e: IOException) {
            e.printStackTrace()
        }

        return ModuleLevelBuilder.ExitCode.ABORT
    }

    private fun runBuild(context: CompileContext, cabal: CabalJspInterface): Boolean {
        context.processMessage(ProgressMessage("Cabal build"))
        val buildProcess = cabal.build()!!
        processCabalOutput(context, collectOutput(buildProcess))
        if (buildProcess.waitFor() != 0) {
            context.processMessage(CompilerMessage("cabal", BuildMessage.Kind.ERROR, "build errors."))
            return true
        }

        return false
    }

    private fun runConfinguration(context: CompileContext, cabal: CabalJspInterface): Boolean {
        context.processMessage(CompilerMessage("cabal", BuildMessage.Kind.INFO, "Start configure"))
        val configureProcess = cabal.configure()!!
        processCabalOutput(context, collectOutput(configureProcess))
        if (configureProcess.waitFor() != 0) {
            context.processMessage(CompilerMessage("cabal", BuildMessage.Kind.ERROR, "configure failed."))
            return true
        }

        context.processMessage(CompilerMessage("cabal", BuildMessage.Kind.INFO, "Start build"))
        return false
    }
    private fun processCabalOutput(context: CompileContext, processOut: Iterator<String>): Unit {
        while (processOut.hasNext()!!)
        {
            val line = processOut.next()
            val warningPrefix = "Warning: "
            if (line.startsWith(warningPrefix)!!)
            {
                val text = line.substring(warningPrefix.length()!!) + "\n" + processOut.next()
                context.processMessage(CompilerMessage("cabal", BuildMessage.Kind.WARNING, text))
            }
            else
            {
                context.processMessage(CompilerMessage("cabal", BuildMessage.Kind.INFO, line))
            }
        }
    }
    private fun collectOutput(process: Process): Iterator<String> {
        val reader = BufferedReader(InputStreamReader(process.getInputStream()!!))
        return object : Iterator<String> {
            var line: String? = null

            public override fun hasNext(): Boolean {
                return fetch() != null
            }

            private fun fetch(): String? {
                if (line == null)
                {
                    try
                    {
                        line = reader.readLine()
                    }
                    catch (e: IOException) {
                        e.printStackTrace()
                    }

                }

                return line
            }

            public override fun next(): String {
                val result = fetch()
                line = null
                return result!!
            }

        }
    }
    private fun getCabalFile(module: JpsModule): File? {
        val url = module.getContentRootsList().getUrls().get(0)
        try {
            for (file in File(URL(url).getFile()!!).listFiles()!!) {
                if (file.getName().endsWith(".cabal")) {
                    return file
                }

            }
        }
        catch (e: MalformedURLException) {
            e.printStackTrace()
        }

        return null
    }
    public override fun getCompilableFileExtensions(): MutableList<String> {
        return ArrayList(Arrays.asList("hs"))
    }
    public override fun toString(): String {
        return getPresentableName()
    }
    public override fun getPresentableName(): String {
        return "Cabal builder"
    }

}
